### DISCUSSION TITLE
GQL - Cart initialization and localization

### DISCUSSION THREAD

**Discussion by henke-89** [swellstores/community#46] [Help] [answered]
Hello, I'm trying out Swell and have run into two things I can't seem to work out and would love some input.

### No 1: Creating a cart
I've been checking out this guide https://developers.swell.is/frontend-api/basic-usage. So I create an empty cart with:

`query getCart($locale: String, $currency: String) {
  cart(_locale: $locale, _currency: $currency) {
    items {
      id
      quantity
      price
      product {
        id
        name
        currency
        price
      }
    }
  }
}`

The `X-Session` token I get back, I save in my local storage.

Now I want to add to this session, by utilizing:

`mutation addToCart($productId: ID, $locale: String, $currency: String) {
  addCartItem(
    _locale: $locale
    _currency: $currency
    input: { productId: $productId, quantity: 1 }
  ) {
    checkoutUrl
    grandTotal
    items {
      id
      quantity
      price
      product {
        id
        name
        currency
      }
    }
  }
}`

Headers are now `auth` + `x-session`. However this does not add the item to the provided session token. Rather it creates a new session, adds the product to it and then returns that new session in the header. Is this is expected? If so, I think I'm misreading the guide. My impression was that I could initiate a new session with an "empty" getCart-request and then use that session?

### No 2: Locales and currencies
I have `EUR`, `USD` and `SEK` set up as currencies in the store, as well as `en-US` and `sv-SE` for locales. Looking at the spec here https://developers.swell.is/frontend-api/carts my understanding is that I could use `_locale` and `_currency` inputs when dealing with the cart to get results in different languages and currencies? So, I try adding something to the cart:

`mutation addToCart($productId: ID, $locale: String, $currency: String) {
  addCartItem(
    _locale: $locale
    _currency: $currency
    input: { productId: $productId, quantity: 1 }
  ) {
    checkoutUrl
    grandTotal
    items {
      id
      quantity
      price
      product {
        id
        name
        currency
      }
    }
  }
}`

This adds the product fine, but returns the cart in the default locale and with the "wrong" price (in base currency) and the currency node as null - regardless of which of my locales and currencies I provide. 

When getting the cart:

`query getCart($locale: String, $currency: String) {
  cart(_locale: $locale, _currency: $currency) {
    items {
      id
      quantity
      price
      product {
        id
        name
        currency
        price
      }
    }
  }
}`

When getting the cart I'm able to get the locale kicking in, but currency is still null - and the price is "wrong" (in the base currency, not adjusted).

Am I misunderstanding how the locales/currencies work?

Any help appreciated! Thanks!

**Comment by rodrigoddalmeida** [comment, marked as answer]
Hi, @henke-89 

On the first topic, yes, that is expected. Our session is a "stateful session", which means some requests can mutate its values and it is safer to sync your local storage with the most recent session coming from requests to Swell. For example, things like locale and currency can be stored in session and change the session value.

**Comment by rodrigoddalmeida** [comment]
As for the second topic, we actually found a bug preventing those query params to be passed in some requests. We have a fix coming up soon, thank you for the heads up!

**Reply by rodrigoddalmeida** [reply]
The bug preventing `$locale | _locale` and `$currency | _currency` to be passed to some cart endpoints has been fixed. Thanks again, @henke-89!

**Reply by henke-89** [reply]
Perfect, thanks! Will have it a go :)

**Reply by henke-89** [reply]
It works...to some extent.

**Mutating products into the cart**
- Locale for returned products is now correct 
- Currency is still not working -> price is always in base currency and currency-node is still null

**Getting the cart**
- Locale is all good
- Price is correct when selected from the product node, but incorrect in the root node
- Currency node is always null

How I'm testing:
- Mutating products with _locale and _currency (not base)
- Querying cart with same _locale and _currency (not base)
Worth mentioning that I'm sending a no-cache header as well so should be hitting updated things

Follow-up question about how things are supposed to work: should I be able to add something to the cart in say EUR and then fetch the cart with currency USD and it should be correct?

**Reply by rodrigoddalmeida** [reply]
The behavior you're seeing is expected, we're aware of the shortcomings of multi-currency in cart responses, as we have been using client-side helper functions (see [Swell.js](https://developers.swell.is/frontend-api/currencies)) to display currency values from those carts. We'll improve the API to make this behavior more consistent.

For more info about how multi-currency works, here's some [additional documentation](https://www.swell.is/help/store-setup/localization/multi-currency#multi-currency).

**Reply by ericingram** [reply]
It shouldn't be necessary to mutate the cart while adding to it every time. Instead, you should set the cart currency by passing the raw currency field without `_`. In Swell.js, this would look like:

```
swell.cart.update({ currency: 'non-base currency' })
```

The meta params `$currency` and `$locale` are for fetching records with those values, but a cart actually has to be priced in said currency using the plain field as above.
